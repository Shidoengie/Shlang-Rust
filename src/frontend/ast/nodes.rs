use crate::frontend::tokens::TokenType;
use crate::spans::*;

use std::collections::*;
use std::fmt::Debug;
#[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
pub enum Precedence {
    Lowest,
    Assign,      // =
    Or,          // or, ||
    And,         // and, &&
    Equality,    // ==, !=
    Comparison,  // <, >, <=, >=
    Nullish,     // ??
    Sum,         // +, -
    Product,     // *, /, %
    Unary,       // -, !
    Call,        // my_func()
    Index,       // my_list[0]
    Constructor, // my_obj{x:10}
    Member,      // my_obj.field
}

/// Helper to get the precedence of a given token type.
impl From<&TokenType> for Precedence {
    fn from(kind: &TokenType) -> Self {
        match kind {
            TokenType::Equal
            | TokenType::QuestionEqual
            | TokenType::PlusEqual
            | TokenType::MinusEqual
            | TokenType::StarEqual
            | TokenType::SlashEqual => Precedence::Assign,
            TokenType::Or | TokenType::DualPipe => Precedence::Or,
            TokenType::And | TokenType::DualAmpersand => Precedence::And,
            TokenType::DoubleEqual | TokenType::BangEqual => Precedence::Equality,
            TokenType::Greater
            | TokenType::GreaterEqual
            | TokenType::Lesser
            | TokenType::LesserEqual => Precedence::Comparison,
            TokenType::DualQuestion => Precedence::Nullish,
            TokenType::Plus | TokenType::Minus => Precedence::Sum,
            TokenType::Slash | TokenType::Star | TokenType::Percent => Precedence::Product,
            TokenType::LParen => Precedence::Call,
            TokenType::LBrace => Precedence::Constructor,
            TokenType::LBracket => Precedence::Index,
            TokenType::Dot => Precedence::Member,
            _ => Precedence::Lowest,
        }
    }
}

/// Helper to convert a token type into its corresponding BinaryOp.
impl From<TokenType> for BinaryOp {
    fn from(kind: TokenType) -> Self {
        match kind {
            TokenType::Plus => BinaryOp::Add,
            TokenType::Minus => BinaryOp::Subtract,
            TokenType::Slash => BinaryOp::Divide,
            TokenType::Star => BinaryOp::Multiply,
            TokenType::Percent => BinaryOp::Modulo,
            TokenType::And | TokenType::DualAmpersand => BinaryOp::And,
            TokenType::Or | TokenType::DualPipe => BinaryOp::Or,
            TokenType::DoubleEqual => BinaryOp::IsEqual,
            TokenType::BangEqual => BinaryOp::IsDifferent,
            TokenType::Greater => BinaryOp::Greater,
            TokenType::Lesser => BinaryOp::Lesser,
            TokenType::GreaterEqual => BinaryOp::GreaterOrEqual,
            TokenType::LesserEqual => BinaryOp::LesserOrEqual,
            TokenType::DualQuestion => BinaryOp::NullCoalescing,
            // This panic should ideally never be reached if the parser logic is correct.
            _ => panic!("Cannot convert token type {:?} to a BinaryOp", kind),
        }
    }
}
#[derive(Clone, Debug, PartialEq)]
pub struct FuncDef {
    pub block: NodeStream,
    pub args: Vec<String>,
    pub captures: bool,
}

#[derive(Clone, PartialEq)]
pub enum Node {
    Null,
    Bool(bool),
    Str(String),
    Float(f64),
    Int(i64),
    BinaryNode(BinaryNode),
    UnaryNode(UnaryNode),
    ResultNode(NodeRef),
    ReturnNode(NodeRef),
    BreakNode,
    ContinueNode,
    Declaration(String, NodeRef),
    Assignment { target: NodeRef, value: NodeRef },
    Variable(String),
    Index { target: NodeRef, index: NodeRef },
    FuncDef(FuncDef),
    ListLit(Vec<NodeSpan>),
    Call(Call),

    Branch(Branch),
    Loop(NodeStream),
    While(While),
    ForLoop(ForLoop),
    DoBlock(NodeStream),
    Constructor(Constructor),
    StructDef(StructDef),
    FieldAccess(FieldAccess),
    DontResult,
}
impl Node {
    pub fn can_result(&self) -> bool {
        if let Self::StructDef(def) = self {
            return def.name.is_none();
        }
        !matches!(
            self.clone(),
            Self::Declaration(_, _)
                | Self::Assignment {
                    target: _,
                    value: _
                }
                | Self::ReturnNode(_)
                | Self::BreakNode
                | Self::ContinueNode
        )
    }
}
impl Debug for Node {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Node::Bool(val) => {
                if *val {
                    write!(f, "true")
                } else {
                    write!(f, "false")
                }
            }

            Node::Int(num) => write!(f, "Int({num})"),
            Node::Float(num) => write!(f, "Float({num})"),
            Node::Str(txt) => write!(f, r#""{txt}""#),
            Node::Variable(val) => write!(f, "Var({val})"),
            Node::Null => f.write_str("Null"),
            Node::BreakNode => f.write_str("Break"),
            Node::ContinueNode => f.write_str("Continue"),
            Node::Declaration(name, expr) => {
                write!(f, "Declare({name} = {expr:?})",)
            }
            Node::Index { target, index } => f
                .debug_struct("Index")
                .field("target", target)
                .field("index", index)
                .finish(),
            Node::ListLit(list) => {
                f.write_str("List")?;
                f.debug_set().entries(list).finish()
            }
            Node::DoBlock(block) => {
                write!(f, "Do")?;
                f.debug_set().entries(block).finish()
            }
            Node::Loop(block) => {
                write!(f, "Loop")?;
                f.debug_set().entries(block).finish()
            }

            Node::ReturnNode(ret) => f.debug_tuple("Return").field(&ret.item).finish(),
            Node::ResultNode(ret) => f.debug_tuple("Result").field(&ret.item).finish(),
            Node::BinaryNode(bin) => write!(f, "{bin:#?}"),
            Node::UnaryNode(node) => write!(f, "{node:#?}"),
            Node::Constructor(node) => write!(f, "{node:#?}"),
            Node::StructDef(node) => write!(f, "{node:#?}"),
            Node::Branch(node) => write!(f, "{node:#?}"),
            Node::ForLoop(node) => write!(f, "{node:#?}"),
            Node::Assignment { target, value } => write!(f, "Assign({target:#?} = {value:#?})"),
            Node::FieldAccess(node) => write!(f, "{node:#?}"),
            Node::While(node) => write!(f, "{node:#?}"),
            Node::Call(node) => write!(f, "{node:#?}"),

            Node::DontResult => f.write_str("DontResult"),

            Node::FuncDef(func) => {
                if func.captures {
                    write!(f, "Closure(")?;
                } else {
                    write!(f, "Function(")?;
                }

                for (index, name) in func.args.iter().enumerate() {
                    f.write_str(name)?;
                    if index + 1 != func.args.len() {
                        write!(f, ", ")?;
                    }
                }
                write!(f, ")")?;

                f.debug_set().entries(&func.block).finish()
            }
        }
    }
}
pub type NodeSpan = Spanned<Node>;
pub type NodeRef = Spanned<Box<Node>>;
impl NodeSpan {
    pub fn wrap_in_result(self) -> Self {
        let span = self.span;
        Node::ResultNode(self.box_item()).to_spanned(span)
    }
}
pub trait IntoNodespan {
    fn to_nodespan(self, span: Span) -> NodeSpan;
}

pub type NodeStream = Vec<Spanned<Node>>;

macro_rules! nodes_from {
    ($($name:ident)*) => {
        $(
            impl ::core::convert::From<$name> for Node {
                fn from(node: $name) -> Self {
                    Self::$name(node)
                }
            }
            impl IntoNodespan for $name {
                fn to_nodespan(self,span:Span) -> NodeSpan {
                    Spanned::new(Node::$name(self.clone()),span)
                }

            }
        )*
    }
}
nodes_from! { FuncDef UnaryNode Constructor StructDef  FieldAccess BinaryNode Call Branch While ForLoop}

#[derive(Clone, Debug, PartialEq)]
pub enum BinaryOp {
    Add,
    Subtract,
    Divide,
    Multiply,
    Modulo,
    And,
    Or,
    IsEqual,
    IsDifferent,
    Greater,
    Lesser,
    GreaterOrEqual,
    LesserOrEqual,
    NullCoalescing,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BinaryNode {
    pub kind: BinaryOp,
    pub left: NodeRef,
    pub right: NodeRef,
}
impl BinaryNode {
    pub fn is(&self, kind: &BinaryOp) -> bool {
        self.kind.eq(kind)
    }
    pub fn isnt(&self, kind: &BinaryOp) -> bool {
        self.kind.ne(kind)
    }
}
#[derive(Clone, Debug, PartialEq)]
pub enum UnaryOp {
    Negative,
    Not,
}
#[derive(Clone, Debug, PartialEq)]
pub struct UnaryNode {
    pub kind: UnaryOp,
    pub object: NodeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Call {
    pub callee: NodeRef,
    pub args: NodeStream,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Field {
    Declaration(String, NodeRef),
    StructDef(StructDef),
}
impl Spanned<Field> {
    pub fn to_node(self) -> NodeSpan {
        match self.item {
            Field::Declaration(name, expr) => Node::Declaration(name, expr).to_spanned(self.span),
            Field::StructDef(decl) => Node::StructDef(decl.clone()).to_spanned(self.span),
        }
    }
}
#[derive(Clone, Debug, PartialEq)]
pub struct FieldAccess {
    pub target: NodeRef,
    pub requested: NodeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructDef {
    pub name: Option<String>,
    pub fields: Vec<Spanned<Field>>,
}
#[derive(Clone, Debug, PartialEq)]

pub struct Constructor {
    pub name: String,
    pub params: HashMap<String, NodeSpan>,
}
#[derive(Clone, Debug, PartialEq)]
pub struct Branch {
    pub condition: NodeRef,
    pub if_block: NodeStream,
    pub else_block: Option<NodeStream>,
}
impl Branch {
    pub fn new_single(condition: NodeSpan, block: NodeStream) -> Self {
        Self {
            condition: condition.box_item(),
            if_block: block,
            else_block: None,
        }
    }
    pub fn new(condition: NodeSpan, if_block: NodeStream, else_block: NodeStream) -> Self {
        Self {
            condition: condition.box_item(),
            if_block,
            else_block: Some(else_block),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct While {
    pub condition: NodeRef,
    pub proc: NodeStream,
}
#[derive(Clone, Debug, PartialEq)]
pub struct ForLoop {
    pub ident: String,
    pub list: NodeRef,
    pub proc: NodeStream,
}
