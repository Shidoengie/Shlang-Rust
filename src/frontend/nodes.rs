use slotmap::{new_key_type, SlotMap};

use crate::backend::scope::Scope;
use crate::spans::*;

use std::collections::*;
use std::fmt::{Debug, Display};
use std::mem;
#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Null,
    Void,
    Num(f64),
    Bool(bool),
    Str(String),
    Function(Function),
    BuiltinFunc(BuiltinFunc),
    Struct(Struct),
    Ref(RefKey),
    List(Vec<Value>),
}
pub trait ValueRepr {
    fn repr(&self) -> String;
}
pub type TypedValue = (Value, Type);
fn list_repr(list: &[Value]) -> String {
    if list.is_empty() {
        return "[]".to_owned();
    }
    let mut out = String::new();
    for val in list {
        out += format!(",{}", val.repr()).as_str();
    }
    out = out.strip_prefix(',').unwrap().to_string();
    out = "[".to_string() + out.as_str() + "]";
    out
}
impl Value {
    pub fn get_type(&self) -> Type {
        match self {
            Value::Bool(_) => Type::Bool,
            Value::BuiltinFunc(_) | Value::Function(_) => Type::Function,
            Value::Null => Type::Null,
            Value::Void => Type::Void,
            Value::Str(_) => Type::Str,
            Value::Num(_) => Type::Num,
            Value::List(_) => Type::List,
            Value::Struct(obj) => Type::Struct(obj.id.clone()),
            Value::Ref(_) => Type::Ref,
        }
    }

    pub fn is_void(&self) -> bool {
        self == &Self::Void
    }
    pub fn matches_typeof(&self, val: &Self) -> bool {
        mem::discriminant(self) == mem::discriminant(val)
    }
}
impl ValueRepr for Value {
    fn repr(&self) -> String {
        match self {
            Self::Num(num) => num.to_string(),
            Self::Bool(cond) => cond.to_string(),
            Self::Str(txt) => format!("\"{txt}\""),
            Self::Null => "null".to_string(),
            Self::Void => "void".to_string(),
            Self::List(list) => list_repr(list),
            Self::Struct(obj) => obj.repr(),
            Self::Function(func) => func.repr(),
            Self::Ref(id) => format!("ref {:?}", id),
            _ => "unnamed".to_string(),
        }
    }
}
impl Display for Value {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let out = match self {
            Self::Num(num) => num.to_string(),
            Self::Bool(cond) => cond.to_string(),
            Self::Str(txt) => txt.to_string(),
            Self::Null => "null".to_string(),
            Self::Void => "void".to_string(),
            Self::List(list) => list_repr(list),
            Self::Struct(obj) => obj.repr(),
            Self::Function(func) => func.repr(),
            Self::Ref(id) => format!("ref{:?}", id),
            _ => "unnamed".to_string(),
        };
        write!(f, "{out}")
    }
}
impl IntoNodespan for Value {
    fn to_nodespan(self, span: Span) -> NodeSpan {
        Spanned::new(Node::Value(self), span)
    }
}
#[derive(Clone, Debug, PartialEq)]
pub struct Struct {
    pub id: Option<String>,
    pub env: Scope,
}
impl ValueRepr for Struct {
    fn repr(&self) -> String {
        let mut buffer = if let Some(name) = &self.id {
            name.to_owned()
        } else {
            String::new()
        };
        buffer += "{ ";
        let vars = &self.env.vars;
        for (k, v) in vars.iter() {
            buffer += k;
            buffer += ":";
            buffer += &v.repr();
            if vars.values().last().unwrap() != v {
                buffer += ", "
            }
        }
        buffer += " }";
        buffer
    }
}
#[derive(Clone, Debug, PartialEq)]
pub struct Function {
    pub block: NodeStream,
    pub args: Vec<String>,
}

impl Function {
    pub fn new(block: NodeStream, args: Vec<String>) -> Self {
        Self { block, args }
    }
}
impl ValueRepr for Function {
    fn repr(&self) -> String {
        let mut buffer = String::from("func(");
        for i in &self.args {
            buffer += i;
            if self.args.last().unwrap() != i {
                buffer += ", "
            }
        }
        buffer += ") ";
        buffer
    }
}
impl From<Function> for Value {
    fn from(x: Function) -> Self {
        Value::Function(x)
    }
}
new_key_type! {
    pub struct RefKey;
}

type FuncPtr = fn(&mut Scope, Vec<Value>, &mut SlotMap<RefKey, Value>) -> Value;
#[derive(Clone)]
pub struct BuiltinFunc {
    pub function: FuncPtr,
    pub arg_size: i16,
    pub id: String,
}
impl Debug for BuiltinFunc {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("BuiltinFunc")
            .field("id", &self.id)
            .field("arg_size", &self.arg_size)
            .finish()
    }
}
impl PartialEq for BuiltinFunc {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}
impl BuiltinFunc {
    pub fn new(id: String, function: FuncPtr, arg_size: i16) -> Self {
        Self {
            function,
            arg_size,
            id,
        }
    }
}
impl From<BuiltinFunc> for Value {
    fn from(x: BuiltinFunc) -> Self {
        Value::BuiltinFunc(x)
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Type {
    Null,
    Void,
    Num,
    Bool,
    Str,
    Function,
    List,
    Struct(Option<String>),
    Ref,
}

impl Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let txt = match self {
            Self::Void => "void",
            Self::Null => "null",
            Self::Function => "func",
            Self::Bool => "bool",
            Self::Num => "num",
            Self::Str => "str",
            Self::List => "list",
            Self::Struct(Some(id)) => id,
            Self::Struct(None) => "struct",
            Self::Ref => "ref",
        };
        write!(f, "{txt}")
    }
}
#[derive(Clone, Debug, PartialEq)]
pub enum Node {
    Value(Value),
    BinaryNode(BinaryNode),
    UnaryNode(UnaryNode),
    ResultNode(Box<NodeSpan>),
    ReturnNode(Box<NodeSpan>),
    BreakNode,
    ContinueNode,
    Declaration(Declaration),
    Assignment(Assignment),
    Variable(String),
    Index {
        target: Box<NodeSpan>,
        index: Box<NodeSpan>,
    },
    ListLit(Vec<NodeSpan>),
    Call(Call),

    Branch(Branch),
    Loop(NodeStream),
    While(While),
    ForLoop(ForLoop),
    DoBlock(NodeStream),
    Constructor(Constructor),
    ConstuctorFunc {
        name: String,
        args: Vec<NodeSpan>,
    },
    StructDef(StructDef),
    FieldAccess(FieldAccess),
    DontResult,
}
impl Node {
    pub fn can_result(&self) -> bool {
        matches!(
            self.clone(),
            Self::Value(_)
                | Self::BinaryNode(_)
                | Self::UnaryNode(_)
                | Self::Variable(_)
                | Self::Call(_)
        )
    }
}

pub type NodeSpan = Spanned<Node>;
pub type NodeRef = Box<Spanned<Node>>;
impl NodeSpan {
    pub fn wrap_in_result(&self) -> Self {
        let value = self.clone();
        Spanned::new(Node::ResultNode(Box::new(value.clone())), value.span)
    }
}
pub trait IntoNodespan {
    fn to_nodespan(self, span: Span) -> NodeSpan;
}

pub type NodeStream = Vec<Spanned<Node>>;

impl From<Value> for Node {
    fn from(x: Value) -> Self {
        Node::Value(x)
    }
}
macro_rules! nodes_from {
    ($($name:ident)*) => {
        $(
            impl ::core::convert::From<$name> for Node {
                fn from(node: $name) -> Self {
                    Self::$name(node)
                }
            }
            impl IntoNodespan for $name {
                fn to_nodespan(self,span:Span) -> NodeSpan {
                    Spanned::new(Node::$name(self.clone()),span)
                }

            }
        )*
    }
}
nodes_from! { UnaryNode Constructor StructDef  FieldAccess BinaryNode Call Assignment Declaration Branch While ForLoop}

#[derive(Clone, Debug, PartialEq)]
pub enum BinaryOp {
    ADD,
    SUBTRACT,
    DIVIDE,
    MULTIPLY,
    MODULO,
    AND,
    OR,
    ISEQUAL,
    ISDIFERENT,
    GREATER,
    LESSER,
    GREATER_EQUAL,
    LESSER_EQUAL,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BinaryNode {
    pub kind: BinaryOp,
    pub left: NodeRef,
    pub right: NodeRef,
}
impl BinaryNode {
    pub fn is(&self, kind: &BinaryOp) -> bool {
        self.kind.eq(kind)
    }
    pub fn isnt(&self, kind: &BinaryOp) -> bool {
        self.kind.ne(kind)
    }
}
#[derive(Clone, Debug, PartialEq)]
pub enum UnaryOp {
    NEGATIVE,
    NOT,
}
#[derive(Clone, Debug, PartialEq)]
pub struct UnaryNode {
    pub kind: UnaryOp,
    pub object: NodeRef,
}
#[derive(Clone, Debug, PartialEq)]
pub struct Declaration {
    pub var_name: String,
    pub value: NodeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Assignment {
    pub target: NodeRef,
    pub value: NodeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Call {
    pub callee: NodeRef,
    pub args: NodeStream,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Field {
    Declaration(Declaration),
    StructDef(StructDef),
}
impl Spanned<Field> {
    pub fn to_node(&self) -> NodeSpan {
        match &self.item {
            Field::Declaration(decl) => NodeSpan::new(Node::Declaration(decl.clone()), self.span),
            Field::StructDef(decl) => NodeSpan::new(Node::StructDef(decl.clone()), self.span),
        }
    }
}
#[derive(Clone, Debug, PartialEq)]
pub struct FieldAccess {
    pub target: NodeRef,
    pub requested: NodeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct StructDef {
    pub name: Option<String>,
    pub fields: Vec<Spanned<Field>>,
}
#[derive(Clone, Debug, PartialEq)]

pub struct Constructor {
    pub name: String,
    pub params: HashMap<String, NodeSpan>,
}
#[derive(Clone, Debug, PartialEq)]
pub struct Branch {
    pub condition: NodeRef,
    pub if_block: NodeStream,
    pub else_block: Option<NodeStream>,
}
impl Branch {
    pub fn new_single(condition: NodeSpan, block: NodeStream) -> Self {
        Self {
            condition: Box::new(condition),
            if_block: block,
            else_block: None,
        }
    }
    pub fn new(condition: NodeSpan, if_block: NodeStream, else_block: NodeStream) -> Self {
        Self {
            condition: Box::new(condition),
            if_block,
            else_block: Some(else_block),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct While {
    pub condition: NodeRef,
    pub proc: NodeStream,
}
#[derive(Clone, Debug, PartialEq)]
pub struct ForLoop {
    pub ident: String,
    pub list: Box<NodeSpan>,
    pub proc: NodeStream,
}
pub type VarMap = HashMap<String, Value>;
