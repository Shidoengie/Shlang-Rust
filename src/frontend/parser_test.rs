use super::nodes::*;
use super::parser::*;

use insta::*;
use std::*;

fn parse_expr(source: &str) -> ParseRes<NodeSpan> {
    let mut parser = Parser::new(source);
    parser.parse_expr()
}
macro_rules! test_func {
    ($func:expr,($($name:ident : $input:expr)*)) => {
        $(
        #[test]
        fn $name () {
            // Perform your desired operation using the input parameter
            assert_debug_snapshot!($func($input));
        }
        )*
    };
}

test_func!(parse_expr,(
parse_nested_paren:"((1+2)+(1+2));"
parse_paren:"(1+2+b);"
expression_operators:"1*2+2-2/2%5 and 10 or 10!=1==2<4<=3>0>=3;"
unary_operator:"!true;"
multiple_unary:"!true + !true;"
unary_operator_paren:"!(0);"
unary_operators:"!0+(-0);"
empty_call:"a();"
call:"a((1+2),true,'1');"
call_as_val:"a(1,1,1)+1;"
call_with_paren:"a((1),1,1);"

assignement:"b = 1;"
single_var_expr_decl:"var a;"
var_expr_decl_with_expr_:"var a = 1+2+b;"
do_block:"do{var a = 1;1+a}"
func_expr:"func(b){1}"
func_decl_expr:"func a (b){1}"
nested_block:"func(b){do{1};}"
nested_block_with_result:"func(b){do{1}1}"
struct_access:"b.a*2"
struct_access_call:"a.b().c"
struct_body:"struct abc{};"
nested_struct:"struct abc{struct dfg{};};"
index:"a[1]"
index_with_call:"a[1]()[2]"
list_literal:"[1,2,3]"
));
