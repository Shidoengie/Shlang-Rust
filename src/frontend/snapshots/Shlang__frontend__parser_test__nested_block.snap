---
source: src/frontend/parser_test.rs
expression: "parse_expr(\"func(b){do{1};}\")"
---
Ok(
    Function(b){
        Do{
            Result(
                Int(1),
            ) at 11:12,
        } at 10:13,
    } at 4:15,
)
